### PR TITLE
BrokerStore<->Zeek table synchronization

### PR DESCRIPTION
**Pull request by 0xxon**
This PR introduces BrokerStore to Zeek table synchronization. This also requires merging the same branch in doc and zeekctl.

A zeek table can be attached to a broker store. After that, changes to the zeek table will automatically be pushed into the backing broker store - and changes to the broker store will be fed back into the zeek table. This should work irrespective of the chosen backend for the broker store.

This basically is the replacement for the old ``&synchronized`` and ``&persistent`` attributes.

When using memory-backed broker stores, the only thing one has to do in zeek is:
```
global table_to_share: table[string] of count &backend=Broker::MEMORY;
```

When using sqlite backed stores, one has to specify a storage location:
```
redef Broker::auto_store_db_directory = "[path]";
global table_to_share: table[string] of count &backend=Broker::SQLITE;
```

When using zeekctl, the storage location is automatically set. Otherwise it is the current directory by default.

This also introduces two more attributes, one which can be used to manually attach broker stores, and another which can be used to disable a few of the type enforcements.

This branch should basically be done. I still have to write more documentation for it, and this relies on https://github.com/zeek/broker/issues/135 being fixed - and currently just uses  a hacky workaround that @jsiwek proposed for that issue.

I would be about feedback/reviews - it is quite possible that there are some loose ends that I lost track of - and cannot find anymore.

### CODE REVIEW

**Comment by jsiwek** [comment]
> this relies on [zeek/broker#135](https://github.com/zeek/broker/issues/135) being fixed - and currently just uses a hacky workaround that @jsiwek proposed for that issue.

I haven't thought of an alternative and would go with what's now in this branch unless Tim has other ideas.

@timwoj this is the change in question:

https://github.com/zeek/zeek/pull/1061/commits/1888d6acaec2183c3cc3130a4d5dec30726cfdca#diff-1c59e2e3f2db8fe6b9f2cda59f909991R922-R926

The Zeek test suites pass even if that `net_update_time()` is completely removed, but guess it could actually help in case nothing else will be updating time (e.g. no packet/event/time activity either ever or just temporarily).

**Comment by timwoj** [comment]
> The Zeek test suites pass even if that net_update_time() is completely removed, but guess it could actually help in case nothing else will be updating time (e.g. no packet/event/time activity either ever or just temporarily).

I added it as part of the ioloop work, because time wasn't moving forward when receiving broker messages but not receiving any other data. Unless something else changed elsewhere, I have to imagine that's still the case. @0xxon tells me we have btests that exercise broker without receiving packet data though, so maybe it's not?

**Comment by 0xxon** [comment]
Added a doc and zeekctl branch with the same name. The documentation is still not quite done; the zeekctl branch should be done.

File: src/broker/Manager.cc [hunk 1]
```diff
@@ -941,6 +998,153 @@ void Manager::Process()
...
+		// it is a table
+		auto zeek_value = data_to_val(data, table->GetType()->Yield().get());
+		if ( ! zeek_value )
+			{
+			reporter->Error("ProcessStoreEvent: could not convert value \"%s\" for key \"%s\" in store \"%s\" while receiving remote insert. This probably means the tables have different types on different nodes.", to_string(data).c_str(), to_string(key).c_str(), insert.store_id().c_str());
+			return;
+			}
+			table->Assign(zeek_key, zeek_value, false);
+		}
+	else if ( auto update = broker::store_event::update::make(msg) )
```
Comment (rsmmr) [inline src/broker/Manager.cc:1052]: `insert` & `update` look very similar, can we factor out the common code?

**Comment by rsmmr** [comment]
one more note: I was wondering if there's a better name than `&backend` for the attribute, but can't come up with anything. Just in case anybody has an idea.

File: src/broker/Manager.h [hunk 2]
```diff
@@ -354,6 +369,13 @@ class Manager : public iosource::IOSource {
 	void ProcessError(broker::error err);
 	void ProcessStoreResponse(StoreHandleVal*, broker::store::response response);
 	void FlushPendingQueries();
+	// Initializes the masters for broker backed zeek tables when using the &backend attribute
+	void InitializeBrokerStoreForwarding();
+	// Check if a broker store is associated to a table on the Zeek side.
+	void CheckForwarding(const std::string& name);
```
[the remark this answers is not in the document]
Comment (timwoj) [reply src/broker/Manager.h:375]: I agree, but `InitForwarding` is already too close to `InitializeBrokerStoreForwarding`.

File: NEWS [hunk 3]
```diff
@@ -55,6 +55,20 @@ New Functionality
   for which no connection state information is available in the core anymore. These
   cases will raise the new ``expired_conn_weird`` event.
 
+- Broker Store table synchronization (Experimental).
+
+  Zeek now supports synchronizing tables/sets across clusters using a backing broker
```
[the remark this answers is not in the document]
Comment (0xxon) [reply NEWS:60]: I will update all of these.

File: scripts/base/frameworks/broker/store.zeek [hunk 4]
```diff
@@ -25,6 +25,16 @@ export {
...
 	const default_clone_mutation_buffer_interval = 2min &redef;
 
+	## If set to true, the current node is the master node for broker stores
+	## backing zeek tables. By default this value will be automatically set to
+	## true in standalone mode, or on the manager node of a cluster. This value
+	## should not typically be changed manually.
+	const auto_store_master = T &redef;
+
+	## The directory used for storing persistent database files when using brokerstore
+	## backed zeek tables.
```
[the remark this answers is not in the document]
Comment (0xxon) [reply scripts/base/frameworks/broker/store.zeek:35]: Similarly - will update all mentions of zeek.

File: scripts/base/frameworks/broker/store.zeek [hunk 5]
```diff
@@ -25,6 +25,16 @@ export {
 	## A negative/zero value indicates to never buffer commands.
 	const default_clone_mutation_buffer_interval = 2min &redef;
 
+	## If set to true, the current node is the master node for broker stores
+	## backing zeek tables. By default this value will be automatically set to
+	## true in standalone mode, or on the manager node of a cluster. This value
+	## should not typically be changed manually.
+	const auto_store_master = T &redef;
```
[the remark this answers is not in the document]
Comment (0xxon) [reply scripts/base/frameworks/broker/store.zeek:32]: I was thinking of automatically forwarding. But - I like table_store_master better - will rename.

File: scripts/base/frameworks/broker/store.zeek [hunk 6]
```diff
@@ -25,6 +25,16 @@ export {
...
+
+	## The directory used for storing persistent database files when using brokerstore
+	## backed zeek tables.
+	const auto_store_db_directory = "." &redef;
```
[the remark this answers is not in the document]
Comment (0xxon) [reply scripts/base/frameworks/broker/store.zeek:36]: Same here - will rename.

File: scripts/base/init-bare.zeek [hunk 7]
```diff
@@ -747,6 +747,7 @@ type script_id: record {
 	enum_constant: bool;	##< True if the identifier is an enum value.
 	option_value: bool;	##< True if the identifier is an option.
 	redefinable: bool;	##< True if the identifier is declared with the :zeek:attr:`&redef` attribute.
+	broker_backend: bool;	##< True if the identifier has a broker backend defined using the  :zeek:attr:`&backend` attribute.
```
[the remark this answers is not in the document]
Comment (0xxon) [reply scripts/base/init-bare.zeek:750]: Actually no - it currently really only sets this to true for backend.

The reason is that this was needed to make &backend work in a clusterized setup. Having this set for &broker_store too would actually break this.

We could introduce another member that gives the value for &broker_store - but I don't currently see a use for this (and there are quite a few table attributes that are not mentioned here).

File: src/Attr.cc [hunk 8]
```diff
@@ -456,6 +458,17 @@ void Attributes::CheckAttr(Attr* a)
...
 
 	case ATTR_EXPIRE_READ:
+		{
+		if ( Find(ATTR_BROKER_STORE) )
+			{
+			Error("&broker_store and &read_expire cannot be used simultaneously");
+			}
+		if ( Find(ATTR_BACKEND) )
+			{
+			Error("&broker_store and &backend cannot be used simultaneously");
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/Attr.cc:468]: Thank you - that was a stupid copy/paste error.

File: src/Attr.cc [hunk 9]
```diff
@@ -456,6 +458,17 @@ void Attributes::CheckAttr(Attr* a)
...
+			}
+		}
+		// fallthrough
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/Attr.cc:471]: No, it won't. The error messages are only raised for &read_expire. After that the typical progessing for all the expire flags has to take place.

File: src/Attr.cc [hunk 10]
```diff
@@ -602,6 +624,101 @@ void Attributes::CheckAttr(Attr* a)
...
+			break;
+			}
+
+		// Temporary since Broker does not support ListVals - and we cannot easily convert to set/vector
+		if ( type->AsTableType()->GetIndexTypes().size() != 1 )
+			{
+			Error("&backend only supports one-element set/table indexes");
+			}
+
+		// Only support atomic types for the moment.
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/Attr.cc:649]: Thanks.

File: src/Val.cc [hunk 11]
```diff
@@ -2166,10 +2298,13 @@ ValPtr TableVal::Remove(const HashKey& k)
 
 	Modified();
 
-	if ( change_func && va )
+	if ( va && ( change_func || ! broker_store.empty() ) )
 		{
 		auto index = table_hash->RecoverVals(k);
-		CallChangeFunc(index.get(), va, ELEMENT_REMOVED);
+		if ( ! broker_store.empty() )
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/Val.cc:2304]: It doesn't - it is never called from the very limited amount of code that deals with brokerstores. I can introduce the flag if you want it - but it just would not be used at the moment (and it seems somewhat unlikely that we would use it in the future).

File: src/broker/Manager.cc [hunk 12]
```diff
@@ -1505,6 +1761,8 @@ bool Manager::CloseStore(const string& name)
 	if ( s == data_stores.end() )
 		return false;
 
+	auto pubid = s->second->store.frontend_id();
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/broker/Manager.cc:1764]: Will remove.

File: src/Val.h [hunk 13]
```diff
@@ -1049,10 +1061,13 @@ class TableVal final : public Val, public notifier::Modifiable {
...
 	enum OnChangeType { ELEMENT_NEW, ELEMENT_CHANGED, ELEMENT_REMOVED, ELEMENT_EXPIRED };
 
-	// Calls &change_func. Does not take ownership of values. (Refs if needed).
-	void CallChangeFunc(const Val* index, const ValPtr& old_value,
+	// Calls &change_func.
+	void CallChangeFunc(const ValPtr& index, const ValPtr& old_value,
 	                    OnChangeType tpe);
 
+	// Sends data on to backing Broker Store
+	void SendToStore(const Val* index, const TableEntryVal* new_entry_val, OnChangeType tpe);
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/Val.h:1069]: I don't think so. There is one place where this is called that only has a const Val* available. I don't think we can get a const Val* converted to a ValPtr without a copy operation.

File: src/broker/Manager.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xxon) [reply src/broker/Manager.h:375]: Yup. I admittedly just was unable to find a great name here. I will change it to InitForwarding for now - unless someone quickly speaks up with a better name.

File: NEWS [hunk 14]
```diff
@@ -55,6 +55,20 @@ New Functionality
...
+  store. The same feature also allows persistent storage of data in tables/sets
+  over Zeek restarts. This feature is implemented using the new ```&backend``` attribute.
+
+  To synchronize a table over a cluster, you can e.g. use:
+
+  ```global t: table[string] of count &backend=Broker::MEMORY;```
+
+  This feature is documented in detail here: FIXME.
```
[the remark this answers is not in the document]
Comment (0xxon) [reply NEWS:68]: Hah. I missed that one. Thanks a lot. I felt like I might have missed one.

File: src/Attr.h [hunk 15]
```diff
@@ -43,6 +42,9 @@ enum AttrTag {
 	ATTR_TYPE_COLUMN,	// for input framework
 	ATTR_TRACKED,	// hidden attribute, tracked by NotifierRegistry
 	ATTR_ON_CHANGE, // for table change tracking
+	ATTR_BROKER_STORE, // for broker-store backed tables
+	ATTR_BROKER_STORE_ALLOW_COMPLEX, // for broker-store backed tables
+	ATTR_BACKEND, // for broker-store backed tabled
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/Attr.h:47]: Thanks :)

File: src/broker/Manager.cc [hunk 16]
```diff
@@ -941,6 +998,153 @@ void Manager::Process()
...
+		{
+		auto storehandle = broker_mgr->LookupStore(insert.store_id());
+		if ( ! storehandle )
+			return;
+
+		auto table = storehandle->forward_to;
+		if ( ! table )
+			return;
+
+		auto key = insert.key();
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/broker/Manager.cc:1013]: God point, thank you.

File: src/broker/Manager.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xxon) [reply src/broker/Manager.h:375]: After thinking about it again - I am going to go with PrepareForwarding.

File: src/broker/Manager.cc [hunk 17]
```diff
@@ -1450,16 +1654,68 @@ StoreHandleVal* Manager::MakeMaster(const string& name, broker::backend type,
...
 
-	// Wait for master to become available/responsive.
-	// Possibly avoids timeouts in scripts during unit tests.
-	handle->store.exists("");
 	return handle;
 	}
 
+void Manager::BrokerStoreToZeekTable(const std::string& name, const StoreHandleVal* handle)
+	{
+	// consider if it might be wise to disable &on_change while filling the table
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/broker/Manager.cc:1671]: Will do.

File: src/broker/Manager.cc (hunk 1 above)
Comment (0xxon) [reply src/broker/Manager.cc:1051]: I refactored them into a separate function. I am not sure if it is really much less complex - feel free to undo this if you don't end up liking it in the end - I put it into a separate commit.

**Comment by 0xxon** [comment]
This is basically done... besides the FreeBSD tests. And I currently have no clue what their problem is. I set up a FreeBSD 12.1 vm (same as the CI) - and the tests pass on it for me.

**Comment by 0xxon** [comment]
Ok, I think this made it pass. Just that now the test infrastructure is unhappy about the external tests :)

**Comment by rsmmr** [comment]
Changes look good to me. If anybody else wants to take a look, please do, otherwise I'll go ahead and merge soon.

**Comment by 0xxon** [comment]
> > * there is no conflict resolution. Simultaneous inserts for the same key will probably lead to a divergent state over the cluster.
> 
> Why's this, and/ could we fix that? The Broker stores should eventually be in sync once everything has been propagated. Won't the tables reflect their content at that point as well?

Sorry - I forgot to answer this one.

While you are right that the broker stores will eventually be in sync once everything has been propagated, this is not true for the tables on the Zeek side. Which is kind of a design decision.

To recap what you probably already know. Broker tables basically stay in sync because everything is ordered through the master. Changes are actually not applied to clones directly; changes only show up locally when the master sends back the change that it applied. Hence, there is a single point that enforces ordering.

This comes with the disadvantage that changes do not immediately appear on the clone side.

For Zeek tables, this seemed like an unacceptable drawback for me. So for Zeek tables that are backed by broker stores, the changes are applied immediately - instead of waiting for the round-trip until the local store has the state applied. I think that this model maps closer to what script authors will expect - you modify a table - and the modification is immediately applied. Just like for all other tables.

However, this also means that it becomes really difficult to globally resolve conflicts - at least with the information that we currently have. Let's assume that we have a node (clone) that sends a couple of updates a second. Broker will send these changes to the master, they will come back - and we will get the change notifications back. In this case this is kind of easy - we know that they originated from us, so we can ignore them.

Now - let's change the scenario and just have two nodes (clones) that change a key simultaneously. For broker this is easy - the manager order is. On the Zeek side - this is harder. At the moment, the change will be applied immediately to the tables. Then the change is sent off to Broker - which will propagate the change to the other node. Each node will ignore its own change - and only apply the change of the other node. Which funnily, in this instance, will mean that each node sees the value that the other node inserted (assuming completely simultaneous insertion).

Resolving this is not easy; it either means that Zeek would have to use the same approach as broker (so - have one choke-point and apply updates delayed), keep track of all changes till they come back from the master to try to figure out the correct resulting ordering - which is not really super easy given the information that we currently have, or check the full state of the broker store against the Zeek store regularly. Which would be possible but expensive.

I hope this is kind of understandable - we could chat about this for a bit sometime :)

In any case - when talking to people it also seemed that this point is not really all that important for most cases that were brought up. The existence of a key - and having one of the values from one of the node was typically seen as sufficient.

File: src/Attr.cc [hunk 18]
```diff
@@ -456,6 +458,17 @@ void Attributes::CheckAttr(Attr* a)
...
 
 	case ATTR_EXPIRE_READ:
+		{
+		if ( Find(ATTR_BROKER_STORE) )
+			{
+			Error("&broker_store and &read_expire cannot be used simultaneously");
+			}
+		if ( Find(ATTR_BACKEND) )
+			{
+			Error("&backend and &broker_store cannot be used simultaneously");
```
Comment (rsmmr) [inline src/Attr.cc:468]: I suppose this should be saying `... and &read_expire`

Comment (0xxon) [reply src/Attr.cc:468]: Yes. I thought I fixed that one, sorry.

**Comment by jsiwek** [comment]
Fairly late to review, but generally wanted to say nice work and that I didn't notice anything wildly concerning to be pesky about.

A possible gotcha may be that the logic in `BrokerStoreToZeekTable()` is all synchronous/blocking requests, but reachable via any `Broker::create_master()` when there's a matching `&broker_store` on some table.  That `Broker::create_master()` could be done somewhere besides `zeek_init()`, causing undesirable "hiccup" (though guess most common usage wouldn't do this anyway).  Might at least be worth a word of warning in documentation if this is all true.

**Pull request merged**